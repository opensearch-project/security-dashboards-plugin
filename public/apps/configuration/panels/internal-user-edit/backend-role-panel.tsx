/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  EuiButton,
  EuiFieldText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiFormRow,
} from '@elastic/eui';
import { isEmpty } from 'lodash';
import React, { Dispatch, Fragment, SetStateAction, useState } from 'react';
import {
  appendElementToArray,
  removeElementFromArray,
  updateElementInArrayHandler,
} from '../../utils/array-state-utils';
import { PanelWithHeader } from '../../utils/panel-with-header';
import { DocLinks, LIMIT_WIDTH_INPUT_CLASS } from '../../constants';

function generateBackendRolesPanels(
  backendRoles: string[],
  setBackendRoles: Dispatch<SetStateAction<string[]>>,
  emptyRoleIndex: number,
  roleEmptyErrorMessage: string,
  setRoleEmptyErrorMessage: Dispatch<SetStateAction<string>>
) {
  const panels = backendRoles.map((backendRole, arrayIndex) => {
    return (
      <Fragment key={`backend-role-${arrayIndex}`}>
        <EuiFlexGroup>
          <EuiFlexItem className={LIMIT_WIDTH_INPUT_CLASS}>
            <EuiFormRow
              label={arrayIndex === 0 ? 'Backend role' : ''}
              error={roleEmptyErrorMessage}
              isInvalid={arrayIndex === emptyRoleIndex && !isEmpty(roleEmptyErrorMessage)}
            >
              <EuiFieldText
                isInvalid={arrayIndex === emptyRoleIndex && !isEmpty(roleEmptyErrorMessage)}
                id={`backend-role-${arrayIndex}`}
                value={backendRole}
                onChange={(e) => {
                  updateElementInArrayHandler(setBackendRoles, [arrayIndex])(e.target.value);
                  setRoleEmptyErrorMessage('');
                }}
                placeholder="Type in backend role"
              />
            </EuiFormRow>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiFormRow hasEmptyLabelSpace={arrayIndex === 0 ? true : false}>
              <EuiButton
                id={`backend-role-delete-${arrayIndex}`}
                color="danger"
                onClick={() => removeElementFromArray(setBackendRoles, [], arrayIndex)}
              >
                Remove
              </EuiButton>
            </EuiFormRow>
          </EuiFlexItem>
        </EuiFlexGroup>
      </Fragment>
    );
  });

  return <>{panels}</>;
}

export function BackendRolePanel(props: {
  state: string[];
  setState: Dispatch<SetStateAction<string[]>>;
}) {
  const { state, setState } = props;
  const [roleEmptyErrorMessage, setRoleEmptyErrorMessage] = useState('');
  const [emptyRoleIndex, setEmptyRoleIndex] = useState(-1);
  // Show one empty row if there is no data.
  if (isEmpty(state)) {
    setState(['']);
  }
  return (
    <PanelWithHeader
      headerText="Backend roles"
      headerSubText="Backend roles are used to map users from external authentication systems, such as LDAP or SAML to OpenSearch security roles."
      helpLink={DocLinks.AccessControlDoc}
      optional
    >
      {generateBackendRolesPanels(
        state,
        setState,
        emptyRoleIndex,
        roleEmptyErrorMessage,
        setRoleEmptyErrorMessage
      )}
      <EuiSpacer />
      <EuiButton
        id="backend-role-add-row"
        onClick={() => {
          if (state.indexOf('') !== -1) {
            setRoleEmptyErrorMessage('Type a backend role before adding a new one');
            setEmptyRoleIndex(state.indexOf(''));
          } else {
            setRoleEmptyErrorMessage('');
            appendElementToArray(setState, [], '');
          }
        }}
      >
        Add another backend role
      </EuiButton>
    </PanelWithHeader>
  );
}
