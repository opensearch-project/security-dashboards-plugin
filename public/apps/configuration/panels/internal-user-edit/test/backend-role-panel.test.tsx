/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiFieldText, EuiFlexGroup } from '@elastic/eui';
import { shallow } from 'enzyme';
import React from 'react';
import {
  appendElementToArray,
  removeElementFromArray,
  updateElementInArrayHandler,
} from '../../../utils/array-state-utils';
import { BackendRolePanel } from '../backend-role-panel';

jest.mock('../../../utils/array-state-utils', () => ({
  appendElementToArray: jest.fn(),
  removeElementFromArray: jest.fn(),
  updateElementInArrayHandler: jest.fn().mockReturnValue(jest.fn()),
  setRoleEmptyErrorMessage: jest.fn(),
}));

describe('User editing - backend role panel', () => {
  const setState = jest.fn();

  const backendRole1 = 'admin';
  const backendRole2 = 'HR';
  const sampleState = [backendRole1, backendRole2];

  describe('BackendRolePanel', () => {
    it('render an empty row when no data', () => {
      shallow(<BackendRolePanel state={[]} setState={setState} />);

      expect(setState).toBeCalledWith(['']);
    });

    it('render data', () => {
      const component = shallow(<BackendRolePanel state={sampleState} setState={setState} />);

      expect(component.find(EuiFlexGroup).length).toBe(2);
      expect(component.find(EuiFieldText).at(0).prop('value')).toBe(backendRole1);
      expect(component.find(EuiFieldText).at(1).prop('value')).toBe(backendRole2);
    });

    it('add row', () => {
      const component = shallow(<BackendRolePanel state={sampleState} setState={setState} />);
      component.find('#backend-role-add-row').simulate('click');

      expect(appendElementToArray).toBeCalledWith(setState, [], '');
    });

    it('change backend role value', () => {
      const component = shallow(<BackendRolePanel state={sampleState} setState={setState} />);
      component.find('#backend-role-0').simulate('change', { target: { value: '' } });

      expect(updateElementInArrayHandler).toBeCalledWith(setState, [0]);
    });

    it('delete row', () => {
      const component = shallow(<BackendRolePanel state={sampleState} setState={setState} />);
      component.find('#backend-role-delete-0').simulate('click');

      expect(removeElementFromArray).toBeCalledWith(setState, [], 0);
    });

    // TODO: Fix the tests for backend role error message
    it('add backend role when the previous role is blank', () => {
      const component = shallow(<BackendRolePanel state={[]} setState={setState} />);
      component.find('#backend-role-add-row').simulate('click');
      component.find('#backend-role-add-row').simulate('click');
      expect(component).toMatchSnapshot();
    });

    it('add backend role when one of the previous roles is blank', () => {
      const component = shallow(<BackendRolePanel state={sampleState} setState={setState} />);
      component.find('#backend-role-0').simulate('change', { target: { value: '' } });
      component.find('#backend-role-add-row').simulate('click');
      expect(component).toMatchSnapshot();
    });
  });
});
