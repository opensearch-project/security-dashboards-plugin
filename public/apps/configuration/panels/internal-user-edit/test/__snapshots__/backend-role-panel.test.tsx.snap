// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User editing - backend role panel BackendRolePanel add backend role when one of the previous roles is blank 1`] = `
<PanelWithHeader
  headerSubText="Backend roles are used to map users from external authentication systems, such as LDAP or SAML to OpenSearch security roles."
  headerText="Backend roles"
  helpLink="https://opensearch.org/docs/latest/security-plugin/access-control/index/"
  optional={true}
>
  <EuiFlexGroup>
    <EuiFlexItem
      className="limit-width-input"
    >
      <EuiFormRow
        describedByIds={Array []}
        display="row"
        error=""
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        isInvalid={false}
        label="Backend role"
        labelType="label"
      >
        <EuiFieldText
          id="backend-role-0"
          isInvalid={false}
          onChange={[Function]}
          placeholder="Type in backend role"
          value="admin"
        />
      </EuiFormRow>
    </EuiFlexItem>
    <EuiFlexItem
      grow={false}
    >
      <EuiFormRow
        describedByIds={Array []}
        display="row"
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={true}
        labelType="label"
      >
        <EuiButton
          color="danger"
          id="backend-role-delete-0"
          onClick={[Function]}
        >
          Remove
        </EuiButton>
      </EuiFormRow>
    </EuiFlexItem>
  </EuiFlexGroup>
  <EuiFlexGroup>
    <EuiFlexItem
      className="limit-width-input"
    >
      <EuiFormRow
        describedByIds={Array []}
        display="row"
        error=""
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        isInvalid={false}
        label=""
        labelType="label"
      >
        <EuiFieldText
          id="backend-role-1"
          isInvalid={false}
          onChange={[Function]}
          placeholder="Type in backend role"
          value="HR"
        />
      </EuiFormRow>
    </EuiFlexItem>
    <EuiFlexItem
      grow={false}
    >
      <EuiFormRow
        describedByIds={Array []}
        display="row"
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        labelType="label"
      >
        <EuiButton
          color="danger"
          id="backend-role-delete-1"
          onClick={[Function]}
        >
          Remove
        </EuiButton>
      </EuiFormRow>
    </EuiFlexItem>
  </EuiFlexGroup>
  <EuiSpacer />
  <EuiButton
    id="backend-role-add-row"
    onClick={[Function]}
  >
    Add another backend role
  </EuiButton>
</PanelWithHeader>
`;

exports[`User editing - backend role panel BackendRolePanel add backend role when the previous role is blank 1`] = `
<PanelWithHeader
  headerSubText="Backend roles are used to map users from external authentication systems, such as LDAP or SAML to OpenSearch security roles."
  headerText="Backend roles"
  helpLink="https://opensearch.org/docs/latest/security-plugin/access-control/index/"
  optional={true}
>
  <EuiSpacer />
  <EuiButton
    id="backend-role-add-row"
    onClick={[Function]}
  >
    Add another backend role
  </EuiButton>
</PanelWithHeader>
`;
